### PR TITLE
feature/JFF-6/proof_sign

### PR DESCRIPTION
**Pull request by maikel-maas**
- Created the callback function;
   - Created the method that accepts the callback function;
   - Created the callback function args object;
   - Created unit tests;

### CODE REVIEW

File: __tests__/VcIssuanceClient.spec.ts [hunk 1]
```diff
@@ -6,53 +9,84 @@ import {
...
-      .withCredentialType('https://imsglobal.github.io/openbadges-specification/ob_v3p0.html#OpenBadgeCredential')
-      .build();
-    const credentialRequest: CredentialRequest = vcIssuanceClient.createCredentialRequest({
-      proof: {
-        proof_type: ProofType.JWT,
-        jwt: 'eyJraWQiOiJkaWQ6ZXhhbXBsZTplYmZlYjFmNzEyZWJjNmYxYzI3NmUxMmVjMjEva2V5cy8xIiwiYWxnIjoiRVMyNTYiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJzNkJoZFJrcXQzIiwiYXVkIjoiaHR0cHM6Ly9zZXJ2ZXIuZXhhbXBsZS5jb20iLCJpYXQiOiIyMDE4LTA5LTE0VDIxOjE5OjEwWiIsIm5vbmNlIjoidFppZ25zbkZicCJ9.ewdkIkPV50iOeBUqMXCC_aZKPxgihac0aW9EkL1nOzM',
-      }
-    });
+    .withCredentialRequestUrl('https://oidc4vci.demo.spruceid.com/credential')
+    .withCredentialRequestUrl('oidc4vci.demo.spruceid.com/credential')
```
Comment (nklomp) [inline __tests__/VcIssuanceClient.spec.ts:57]: Why does it have 2 requestUrls of which one is incorrect?

File: src/main/VcIssuanceClient.ts [hunk 2]
```diff
@@ -55,6 +66,25 @@ export class VcIssuanceClient {
...
+  public async createProofOfPossession(opts: {
+    jwtSignerCallback: JWTSignerCallback;
+    jwtSignerArgs: JWTSignerArgs;
+  }): Promise<ProofOfPossession> {
+    if (!opts.jwtSignerCallback || !opts.jwtSignerArgs) {
+      throw new Error(BAD_PARAMS);
+    }
+    return {
+      proof_type: ProofType.JWT,
+      jwt: await opts.jwtSignerCallback(opts.jwtSignerArgs),
```
Comment (nklomp) [inline src/main/VcIssuanceClient.ts:84]: We should check whether the callback returns a valid jwt

**Review by nklomp — CHANGES_REQUESTED** [review]
Some remarks

File: src/main/types/VCIssuance.types.ts [hunk 3]
```diff
@@ -61,3 +63,36 @@ export interface ProofOfPossession {
...
+  kid?: string; // CONDITIONAL. JWT header containing the key ID. If the Credential shall be bound to a DID, the kid refers to a DID URL which identifies a particular key in the DID Document that the Credential shall be bound to. MUST NOT be present if jwk or x5c is present.
+  jwk?: JWK; // CONDITIONAL. JWT header containing the key material the new Credential shall be bound to. MUST NOT be present if kid or x5c is present.
+  x5c?: string[]; // CONDITIONAL. JWT header containing a certificate or certificate chain corresponding to the key used to sign the JWT. This element may be used to convey a key attestation. In such a case, the actual key certificate will contain attributes related to the key properties. MUST NOT be present if kid or jwk is present.
+}
+
+export interface JWTPayload {
+  iss: string; // REQUIRED (string). The value of this claim MUST be the client_id of the client making the credential request.
+  aud: string; // REQUIRED (string). The value of this claim MUST be the issuer URL of credential issuer.
+  iat: number; // REQUIRED (number). The value of this claim MUST be the time at which the proof was issued using the syntax defined in [RFC7519].
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply src/main/types/VCIssuance.types.ts]: Fixed

File: __tests__/VcIssuanceClient.spec.ts (hunk 1 above)
Comment (maikel-maas) [reply __tests__/VcIssuanceClient.spec.ts]: Merge problem, fixed

File: src/main/types/VCIssuance.types.ts [hunk 4]
```diff
@@ -61,3 +63,36 @@ export interface ProofOfPossession {
...
+export interface JWTHeader {
+  alg: string; // REQUIRED by the JWT signer
+  kid?: string; // CONDITIONAL. JWT header containing the key ID. If the Credential shall be bound to a DID, the kid refers to a DID URL which identifies a particular key in the DID Document that the Credential shall be bound to. MUST NOT be present if jwk or x5c is present.
+  jwk?: JWK; // CONDITIONAL. JWT header containing the key material the new Credential shall be bound to. MUST NOT be present if kid or x5c is present.
+  x5c?: string[]; // CONDITIONAL. JWT header containing a certificate or certificate chain corresponding to the key used to sign the JWT. This element may be used to convey a key attestation. In such a case, the actual key certificate will contain attributes related to the key properties. MUST NOT be present if kid or jwk is present.
+}
+
+export interface JWTPayload {
+  iss: string; // REQUIRED (string). The value of this claim MUST be the client_id of the client making the credential request.
+  aud: string; // REQUIRED (string). The value of this claim MUST be the issuer URL of credential issuer.
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply src/main/types/VCIssuance.types.ts]: Fixed

File: src/main/types/VCIssuance.types.ts [hunk 5]
```diff
@@ -61,3 +63,36 @@ export interface ProofOfPossession {
...
+  kty?: string;
+  crv?: string;
+  x?: string;
+  y?: string;
+  e?: string;
+  n?: string;
+}
+
+export interface JWTHeader {
+  alg: string; // REQUIRED by the JWT signer
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply src/main/types/VCIssuance.types.ts]: Fixed

File: README.md [hunk 6]
```diff
@@ -46,105 +46,120 @@ export interface CredentialResponseError {
...
-//         op_state: 'eyJhbGciOiJSU0Et...FYUaBy'
-//     },
-//     {
-//       issuer: 'https://server.example1.com',
-//           credential_type: 'https://did.example1.org/healthCard',
-//         op_state: 'eyJhbGciOiJSU0Et...FYUaBy'
-//     },
-// ]
+const jwtArgs: JWTSignerArgs = {
+  header: {
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply README.md:163]: Fixed

File: src/main/VcIssuanceClient.ts [hunk 7]
```diff
@@ -71,20 +73,46 @@ export class VcIssuanceClient {
...
+    jwtVerifyCallback: JWTVerifyCallback;
   }): Promise<ProofOfPossession> {
-    if (!opts.jwtSignerCallback || !opts.jwtSignerArgs) {
+    if (!opts.jwtSignerCallback || !opts.jwtSignerArgs || !opts.jwtVerifyCallback) {
       throw new Error(BAD_PARAMS);
     }
+    const signerArgs = this.setJWSDefaults(opts.jwtSignerArgs);
+    const jwt = await opts.jwtSignerCallback(signerArgs);
+    try {
+      const algorithm = opts.jwtSignerArgs.header.alg;
```
Comment (nklomp) [inline src/main/VcIssuanceClient.ts:89]: A simple check whether it started with 'ey' contained 2 dots would have been enough, but this is nicer. Although this does force the implementor to provide a jwtVerifyCallback implementation.

Maybe make that optional. If present use the verify callback to verify the JWT. IF not present use the simple version checking for 3 parts seperated by dots and starting with ey

**Review by nklomp — CHANGES_REQUESTED** [review]
Nice. Only one remark about the verifyCallback. I think we can make that optional and do a poormans version of it when not present. That would make it easer for an implementor

**Comment by nklomp** [comment]
BTW, PR has conflicts

File: src/main/VcIssuanceClient.ts (hunk 7 above)
Comment (maikel-maas) [reply src/main/VcIssuanceClient.ts]: Fixed!

**Review by nklomp — APPROVED** [review]
LGTM

**Review by nklomp — APPROVED** [review]
LGTM

**Pull request merged**